Exercise_recursive/recursive-array.sh

```diff
@@ -0,0 +1,97 @@
+#!/bin/bash
+
+# This script submits an array job on a single node with the 
+# highest array index job launching the next set of job recursively.
+
+# Usage:
+#   sbatch recursive-array.sh multiplier-initial-value jobs-max-limit
+#   For example:
+#   sbatch recursive-array.sh 1 60000
+
+#SBATCH --partition=reservation                                                                                                                                                                                    
+#SBATCH --reservation=fall_training_cpu_2022
+#SBATCH --nodes=1
+#SBATCH --cpus-per-task=1
+#SBATCH --time=01:00:00
+#SBATCH --mem-per-cpu=100M
+#SBATCH --array=[0-4]
+
+# Use a regular partition, such as express, when not in a training                                                                                                                                                                  
+# session                                                                                                                                                                                                                   
+##SBATCH --partition=express
+
+# Does not redirect output or error to files.
+
+# Check if max limit for jobs is provided by the user or not
+JOBLIMIT=$2
+if [ -z "$JOBLIMIT" ] || [ $# -lt 2 ]
+then
+    echo "This script requires total number of jobs needed to be run as its 2nd command line argument"
+    exit
+fi
+
+echo "JOBLIMIT is $JOBLIMIT"
+
+if [ $JOBLIMIT -lt $SLURM_ARRAY_TASK_COUNT ]
+then
+    echo "The max limit for jobs needs to at least match the total number of array jobs being launched"
+    exit
+fi
+
+# To keep track of Iteration, create a counter
+MULTIPLIER=$1
+if [ -z "$MULTIPLIER" ]
+then
+  MULTIPLIER=1
+fi
+
+WORKDIR=$PWD
+echo "Main directory for recursive-array script is: $WORKDIR"
+
+# Create a directory corresponding to each iteration,
+# Slurm Job ID, & Slurm Array Index.
+DIRNAME=d${MULTIPLIER}-${SLURM_ARRAY_JOB_ID}-${SLURM_ARRAY_TASK_ID}
+mkdir $DIRNAME
+
+# Switch to that directory, capture output, and 
+# execute your science there
+echo "Entering directory $DIRNAME" 
+pushd $DIRNAME
+echo "In $DIRNAME" >> screenlog.out
+
+echo "Multiplier is $MULTIPLIER" >> screenlog.out
+
+# Determine my task-id
+MY_TASK=$((MULTIPLIER * 10 + SLURM_ARRAY_TASK_ID))
+
+echo "Job array ID: $SLURM_ARRAY_JOB_ID , sub-job $SLURM_ARRAY_TASK_ID is running!" >> screenlog.out
+echo "Highest job array index value is $SLURM_ARRAY_TASK_MAX" >> screenlog.out
+echo "Number of tasks in array job is $SLURM_ARRAY_TASK_COUNT" >> screenlog.out
+echo "MY_TASK is $MY_TASK" >> screenlog.out
+
+# Science happens here
+python $WORKDIR/test_function_script.py > test-array-express-$SLURM_ARRAY_JOB_ID-$SLURM_ARRAY_TASK_ID.out
```

Comment:
I see python is called but I can't see a python/anaconda module loaded in the script. is it just using the default 2.7.5 version on the system?

Reply:
@marianalevineu , yes, just using the default version. This is taken from your HPC array example. It's a simple python program, so no need for an anaconda module, I guess.

Reply:
@manasvita, can we use `python3` instead, just to get in the habit?  Python 2.x is long-deprecated.

Reply:
@dhml, sure, just did. Please check and let me know if this looks okay to you. Thanks.

Reply:
@manasvita, looks like it's still `python` being invoked in this script.